### PR TITLE
cmd/snap, daemon: add debug command for getting OnDiskVolume dump

### PR DESCRIPTION
**Pull request by anonymouse64**
Simple utility for getting the OnDiskVolume's for all disks on a device for debugging purposes. It's not super clean and contains lots of `null` since we don't have any json tags defined, but IMHO for a debug thing like this we don't need it.

Example output:

```json
[
    {
        "Structure": [
            {
                "VolumeName": "",
                "Name": "BIOS Boot",
                "Label": "",
                "Offset": null,
                "OffsetWrite": null,
                "Type": "21686148-6449-6E6F-744E-656564454649",
                "Role": "",
                "ID": "b2e891ee-b971-4a2b-b874-694bbf9b821a",
                "Filesystem": "",
                "Content": null,
                "Update": {
                    "Edition": 0,
                    "Preserve": null
                },
                "StartOffset": 1048576,
                "AbsoluteOffsetWrite": null,
                "Index": 1,
                "LaidOutContent": null,
                "ResolvedContent": null,
                "Node": "/dev/sda1",
                "Size": 1048576
            },
            {
                "VolumeName": "",
                "Name": "EFI System",
                "Label": "system-boot",
                "Offset": null,
                "OffsetWrite": null,
                "Type": "C12A7328-F81F-11D2-BA4B-00A0C93EC93B",
                "Role": "",
                "ID": "a87e9dcb-b1e1-4eab-89cf-1c2fc057b038",
                "Filesystem": "vfat",
                "Content": null,
                "Update": {
                    "Edition": 0,
                    "Preserve": null
                },
                "StartOffset": 2097152,
                "AbsoluteOffsetWrite": null,
                "Index": 2,
                "LaidOutContent": null,
                "ResolvedContent": null,
                "Node": "/dev/sda2",
                "Size": 52428800
            },
            {
                "VolumeName": "",
                "Name": "writable",
                "Label": "writable",
                "Offset": null,
                "OffsetWrite": null,
                "Type": "0FC63DAF-8483-4772-8E79-3D69D8477DE4",
                "Role": "",
                "ID": "cba2b8b3-c2e4-4e51-9a57-d35041b7bf9a",
                "Filesystem": "ext4",
                "Content": null,
                "Update": {
                    "Edition": 0,
                    "Preserve": null
                },
                "StartOffset": 54525952,
                "AbsoluteOffsetWrite": null,
                "Index": 3,
                "LaidOutContent": null,
                "ResolvedContent": null,
                "Node": "/dev/sda3",
                "Size": 10682875392
            }
        ],
        "ID": "2a9b0671-4597-433b-b3ad-be99950e8c5e",
        "Device": "/dev/sda",
        "Schema": "gpt",
        "Size": 10737418240,
        "UsableSectorsEnd": 20971487,
        "SectorSize": 512
    }
]
```

### CODE REVIEW

File: cmd/snap/cmd_debug_disks.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+	"github.com/snapcore/snapd/gadget"
+)
+
+type cmdDiskMapping struct {
+	clientMixin
+}
+
+func init() {
+	cmd := addDebugCommand("disks",
+		"(internal) obtain all on disk volumes",
```
Comment (mardy) [inline cmd/snap/cmd_debug_disks.go:36]: Is the "on" a leftover, or is a work ("info" maybe?) missing?

**Review by mardy — APPROVED** [review]
LGTM, the help description is not very clear, but this is nitpicking.

File: cmd/snap/cmd_debug_disks.go (hunk 1 above)
Comment (mvo5) [reply cmd/snap/cmd_debug_disks.go:36]: (bike-shed^9) maybe `(internal) obtain all on-disk volumes information as yaml` ?

**Review by mvo5 — APPROVED** [review]
Thanks, this looks fine to me

File: daemon/api_debug.go [hunk 2]
```diff
@@ -307,6 +327,8 @@ func getDebug(c *Command, r *http.Request, user *auth.UserState) Response {
 		return getChangeTimings(st, chgID, ensureTag, startupTag, all == "true")
 	case "seeding":
 		return getSeedingInfo(st)
+	case "disks":
+		return getDisks(st)
```
Comment (mvo5) [inline daemon/api_debug.go:331]: One (probably silly) question: Is this available to normal users and if so, could we somehow disclose information that normal users would not get via this command (i.e. should we restrict it to "root")?

File: cmd/snap/cmd_debug_disks.go (hunk 1 above)
Comment (anonymouse64) [reply cmd/snap/cmd_debug_disks.go:36]: @mvo5's suggestion looks good to me

File: daemon/api_debug.go (hunk 2 above)
Comment (anonymouse64) [reply daemon/api_debug.go:331]: udevadm on my Ubuntu Core device is accessible to normal users, so I don't see any extra information we would be leaking here:

```
ubuntu@ubuntu:~$ udevadm info --query property /dev/vda1
DEVPATH=/devices/pci0000:00/0000:00:03.0/virtio1/block/vda/vda1
DEVNAME=/dev/vda1
DEVTYPE=partition
PARTN=1
PARTNAME=BIOS Boot
MAJOR=252
MINOR=1
SUBSYSTEM=block
USEC_INITIALIZED=5024163
ID_PATH=pci-0000:00:03.0
ID_PATH_TAG=pci-0000_00_03_0
ID_PART_TABLE_UUID=734398f2-f5bd-4d9f-b429-5532e772da10
ID_PART_TABLE_TYPE=gpt
ID_PART_ENTRY_SCHEME=gpt
ID_PART_ENTRY_NAME=BIOS\x20Boot
ID_PART_ENTRY_UUID=107e0b6e-6315-4675-8d41-28322a8f6da6
ID_PART_ENTRY_TYPE=21686148-6449-6e6f-744e-656564454649
ID_PART_ENTRY_NUMBER=1
ID_PART_ENTRY_OFFSET=2048
ID_PART_ENTRY_SIZE=2048
ID_PART_ENTRY_DISK=252:0
DEVLINKS=/dev/disk/by-partlabel/BIOS\x20Boot /dev/disk/by-path/virtio-pci-0000:00:03.0-part1 /dev/disk/by-path/pci-0000:00:03.0-part1 /dev/disk/by-partuuid/107e0b6e-6315-4675-8d41-28322a8f6da6
TAGS=:systemd:
ubuntu@ubuntu:~$ id
uid=1000(ubuntu) gid=1000(ubuntu) groups=1000(ubuntu)
```

File: cmd/snap/cmd_debug_disks.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+ */
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/jessevdk/go-flags"
+	"github.com/snapcore/snapd/gadget"
```
Comment (bboozzoo) [inline cmd/snap/cmd_debug_disks.go:27]: linter complains about missing empty line between snapd imports and go-flags

File: cmd/snap/cmd_debug_disks.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+func (x *cmdDiskMapping) Execute(args []string) error {
+	if len(args) > 0 {
+		return ErrExtraArgs
+	}
+	resp := []gadget.OnDiskVolume{}
+	if err := x.client.DebugGet("disks", &resp, nil); err != nil {
+		return err
+	}
+	b, err := json.Marshal(resp)
```
Comment (bboozzoo) [inline cmd/snap/cmd_debug_disks.go:52]: hm the help message states `... as yaml`?

File: cmd/snap/cmd_debug_disks.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+	"github.com/jessevdk/go-flags"
+	"github.com/snapcore/snapd/gadget"
+)
+
+type cmdDiskMapping struct {
+	clientMixin
+}
+
+func init() {
+	cmd := addDebugCommand("disks",
```
Comment (bboozzoo) [inline cmd/snap/cmd_debug_disks.go:35]: One more thing that would probably be useful is a debug command to print out the disk layout according to the gadget, eg `snap debug show-layout /snap/my-gadget/current` and then one could compare if things match up.

File: daemon/api_debug.go [hunk 6]
```diff
@@ -307,6 +327,8 @@ func getDebug(c *Command, r *http.Request, user *auth.UserState) Response {
 		return getChangeTimings(st, chgID, ensureTag, startupTag, all == "true")
 	case "seeding":
 		return getSeedingInfo(st)
+	case "disks":
```
Comment (bboozzoo) [inline daemon/api_debug.go:330]: This is implemented via the debug endpoint to avoid importing the gadget package in snap?

**Review by bboozzoo — APPROVED** [review]

File: daemon/api_debug.go (hunk 2 above)
Comment (anonymouse64) [reply daemon/api_debug.go:331]: ah actually trying to implement Maciej's (implicit) suggestion about just calling the disks / gadget functions directly from the snap command led me to find that the one thing we cannot do as a normal user is use blockdev to get the sector size:

```
$ ./snap-exe debug disks
error: cannot get on disk volume for device /dev/nvme0n1: blockdev: cannot open /dev/nvme0n1: Permission denied
```

however I still stand behind my point that the sector size of a disk is not a meaningfully useful privilege escalation :-)

File: cmd/snap/cmd_debug_disks.go (hunk 5 above)
Comment (anonymouse64) [reply cmd/snap/cmd_debug_disks.go:35]: yes I have that debug command too in a different branch waiting on some pre-reqs first

File: cmd/snap/cmd_debug_disks.go (hunk 4 above)
Comment (anonymouse64) [reply cmd/snap/cmd_debug_disks.go:52]: ah yes that help message should be JSON

File: cmd/snap/cmd_debug_disks.go [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+ *
+ */
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/snapcore/snapd/gadget"
```
Comment (bboozzoo) [inline cmd/snap/cmd_debug_disks.go]: Hm still something off with the import order. I'll push a patch

Comment (anonymouse64) [reply cmd/snap/cmd_debug_disks.go]: thanks for fixing this

**Pull request merged**